crates/bevy_pbr/src/bundle.rs

```diff
@@ -75,6 +75,19 @@ pub struct PointLightBundle {
     pub visibility: Visibility,
 }
 
+/// A component bundle for spot light entities
+#[derive(Debug, Bundle, Default)]
+pub struct SpotLightBundle {
+    pub point_light: PointLight,
```

Comment:
My biggest issue with 4 is that we can't really know what fields are common across all future lights. For example, the "common" fields I listed like color + shadow info may not apply to other future light types, which might get their color from a texture, or might emit light in a "volume" or something AO-style that doesn't cast shadows.

And more important to this pr, the "intersection" between Point+Spot lights probably won't be the same as the intersection between all lights. So in a way it feels orthogonal to the problem at hand. Path 3 feels like the safer middle ground because it doesn't risk needing to move things around in the future because it more strictly defines the intersection. That being said, we should consider adding a Light marker component at the very least, just to let people easily query all lights.

Reply:
And then we could move "common across all lights" fields into the Light struct as things stabilize + consensus emerges.

Reply:
Hmm, right! I would be okay with a model of Path 3 -> maybe Path 4 as things stabilize. Path 3 is definitely my clear second choice.

Reply:
Another reason to favor Path 1 is simplicity for the common case. The "common" use case is a user wants to query a specific light of a specific type (ex: SpotLight) and access one or more fields on it. The next most common case will probably be the same, but querying all lights of a specific type.

```rust
fn system(mut spot_lights: Query<&mut SpotLight>) {
  for mut spot_light in spot_lights.iter_mut() {
    spot_light.color = FOO;
    spot_light.inner_angle = BAR;
  } 
}
```

If we split out "common" fields across multiple structs, the user then needs to query the Light component _and_ the SpotLight component. Idk if the "theoretical conceptual win" of split components outweighs the cost of complicating ~99.5% common case. 


```rust
fn system(mut spot_lights: Query<(&mut SpotLight, &mut Light)>) {
  for (mut light, mut spot_light) in spot_lights.iter_mut()
    light.color = FOO;
    spot_light.inner_angle = SOMETHING;
  } 
}
```

Reply:
I think for this type of split, we should build additional abstractions as users have cases that demand them. So far, I have never encountered something that would benefit from "commonizing" PointLight and SpotLight fields, but I have encountered many cases that would lose clarity as a result. That gives me pause.

Reply:
I'm not technically qualified to add to this discussion but I feel this is an area where your going to have artists, developers and designers all give you different opinions. Also I'd like to throw you a link to this book (3rd edition is free online),  [https://pbr-book.org/](https://pbr-book.org/3ed-2018/Light_Sources) for your research which has an interesting section on lighting.

Reply:
I think I would be in favor of Option 1, but I wanted to share some things:

First, how you would have to query lights if you wanted all lights and went with Option 1:
```rust
#[derive(Component)]
struct LightMarker {}

// note - have to use a ParamSet because an entity could be a PointLight and a SpotLight
fn query_all_lights(
    mut set: ParamSet<(
        Query<&mut PointLight, With<LightMarker>>,
        Query<&mut SpotLight, With<LightMarker>>,
        Query<&mut AmbientLight, With<LightMarker>>,
        Query<&mut DirectionalLight, With<LightMarker>>,
        // expand ParamSet for each type of new light
    )>,
) {
    // do stuff here
} 
```

I toyed around with putting all lights in an enum:

```rust
struct PointLightStruct {
    pub color: Color,
    pub intensity: f32,
    pub range: f32,
    /* other common fields here */
}

struct SpotLightStruct {
    pub color: Color,
    pub intensity: f32,
    pub range: f32,
    /* other common fields here */
    pub inner_angle: f32,
    pub outer_angle: f32,
}

struct AmbientLightStruct {
    // stuff
}

struct DirectionalLightStruct {
    // stuff
}

enum LightEnum {
    PointLight(PointLightStruct),
    SpotLight(SpotLightStruct),
    AmbientLight(AmbientLightStruct),
    DirectionalLight(DirectionalLightStruct),
}

struct Light(pub LightEnum);
```

which would make it simpler to query all lights at the cost of making it more difficult to query just a certain kind of light.

Reply:
Personally, I'm in favor of path 1. As you said, most operation will be done on specific types and it's really hard to say what actually is common to all lights. It's also easier to work with if you don't have to look at multiple structs just to configure a point light.

Reply:
i think i favour option 2.

from an editor perspective whatever we do the fields should be flattened and these structs should be transparent / invisible, or at worst separate headings in an editor tab - personally i don't think that is a big problem. it does make life more awkward from a code-first perspective, but it definitely feels more like the standard ECS approach to use multiple small components, so it's not a problem unique to this area.

filament groups point and spot lights separately to directional lights as "punctual" lights, meaning lights with a physical location. from a technical / implementation perspective there is shared logic (particularly for clustering), so i think there's value to having `PunctualLight` (or `PositionalLight` if punctual is too obtuse) shared between point and spot bundles. 

i am not sure of the value of the `PointLight` marker (and not sure what we would/should do if we have both `PointLight` and `SpotLight` on the same entity). with a shared component, a point light query would be equivalent to `Query<(&Light, &PunctualLight>), Without<SpotLight>>` but i guess that may change in future.

i am not convinced by a common light struct either... the shadow params and brightness are differently specified / interpreted for directional lights and ambient lights, and i don't really see a way to reconcile them.
for brightness, ambient light and directional light have an explicit capped range where the punctual lights are uncapped. for shadows, ambient lights can't have any shadows at all, and directional lights have an additional projection field (i am not so sure this couldn't be reconciled though).

so I would suggest:
```rust
struct PunctualLight {
    pub color: Color,
    pub intensity: f32,
    pub range: f32,
    pub radius: f32,
    pub shadows_enabled: bool,
    pub shadow_depth_bias: f32,
    pub shadow_normal_bias: f32,
}

struct SpotLight {
    pub range: f32,
    pub radius: f32,
    pub inner_angle: f32,
    pub outer_angle: f32,
}
```

and bundles 

```rust
struct PointLightBundle {
    TransformBundle,
    PunctualLight,
    // implementation details (frusta etc)
}

struct SpotLightBundle {
    TransformBundle,
    PunctualLight,
    SpotLight,
   // implementation details (frusta etc)
}}
```

and leave ambient and directional as is.

Reply:
I think for now my vote is Path 1. It is the simplest and clearest user experience and only comes at the cost of slight duplication at the component level. All other major ecosystem players present these light types as "fully separate" types and I would like to default to that in our code-first apis. This approach is also consistent with the pattern we've used for lights up until this point. We can have the "how do we abstract over light types" conversation later ... that deserves its own discussion on its own time. Not something I want to rush into for 0.8.

Port to Path 1 and I'll merge this asap. Feel free to continue the conversation if you'd like though. But given that 0.8 is impending and I hear @robtfm will be out starting saturday, I'll probably port this to Path 1 myself if we haven't reached a conclusion / final impl by the start of next week (or I might try to recruit someone else to do the port :smile: )

Reply:
For what it’s worth, I agree with it being weird that a SpotLightBundle has a PointLight component if we have a PointLightBundle. I was mostly thinking about how they are constructed when I suggested a separate bundle to Rob.

I don’t like doing too much abstraction/generalisation until it is clear that it will be stable and the situation is known, or there is a clear benefit to it and you can still punch down to the lower levels and build whatever you want such that the abstraction is just a convenience for a common case. As such, I was leaning toward path 1 after the options were presented. I don’t think there is a significant negative to it. We can convert to whatever data format we like in the render app for the GPU to use. And I was most concerned about not trying to interpret the intensity field between different light types as being the same because they have different units and behaviours.

Reply:
i've done option 1 for the user-facing part only - there's no real code duplication since we can still use shared `ExtractedPointLight` and `GpuPointLight` structs. those (and the wgsl PointLight struct) should maybe be renamed to make the shared use clearer. i will do that later today if you want, or it should be easy for someone else to pick up.

Reply:
I'll let @cart decide on the naming there. I'd lean toward `ExtractedPunctualLight` or `ExtractedPositionalLight` or something. Or just leave it as it is.

One more thing that I meant to mention - the directional light shadow projection will go away when cascaded shadow maps are added. I think we maybe added the projection as part of the light so it wouldn't get picked up by some other systems that look for `OrthographicProjection`. However, I've made a specific component to house the orthographic projections used for cascades as that makes the most sense there.

Reply:
Cool cool! Treating spotlights as pointlights (terminology-wise) in the renderer internals _is_ slightly confusing, but not so much that I think we should block on it. Using a new term like "punctual" might help, but we can iterate. I think this is good to go for now!